### PR TITLE
Ledger: add APIs for getting allocation and distribution by ID

### PR DESCRIPTION
**Pull request by blueridger**
# Description
This adds APIs for looking up an allocation or a distribution by ID by adding lookup maps as properties on the Ledger and filing those maps during the parsing of the DISTRIBUTE_GRAIN actions in the ledger log. Access to a distribution lets us compare all related allocations and see the timestamp. Access to an allocation gives us information about policy type and lets us compare all related receipts (identity-level payouts).

This is a co-requisite for https://github.com/sourcecred/sourcecred/pull/2534 to surface the allocationId and distributionId, which will allow future development of features that provide transparent reporting on how grain is being distributed, such as the new Explorer Home page.

# Test Plan
`yarn test` passes, unit tests added for full coverage

### CODE REVIEW

File: src/core/ledger/ledger.js [hunk 1]
```diff
@@ -159,6 +165,47 @@ export class Ledger {
...
+  distributionByAllocationId(allocationId: UuidT): Distribution {
+    for (const distribution of this._distributions.values()) {
+      if (
+        distribution.allocations.find(
+          (allocation) => allocation.id === allocationId
+        )
+      )
+        return distribution;
+    }
+    throw new Error(`no Distribution for allocation id: ${allocationId}`);
```
Comment (decentralion) [inline src/core/ledger/ledger.js:206]: I think we should invest upfront in making this performant, by adding a `_allocationIdToDistribution: Map<AllocationId, DistributionId>` field to the ledger.

**Review by decentralion — COMMENTED** [review]
Looks pretty good. In addition to the performance improvement suggested, we should add an API method which returns an iterator over all the allocations and another which returns an iterator over the distributions. We should iterate over them in logical order (i.e. same order as in the event log)

File: src/core/ledger/ledger.test.js [hunk 2]
```diff
@@ -847,21 +879,118 @@ describe("core/ledger/ledger", () => {
...
+            paid: g("17"),
+            active: true,
+          };
+          expect(ledger.accounts()).toEqual([ac1, ac2]);
+        });
+
+        it("should record the allocations and distributions", () => {
+          expect(ledger.allocation(allocation1.id)).toEqual(allocation1);
+          expect(ledger.allocation(allocation2.id)).toEqual(allocation2);
+          expect(ledger.distribution(distribution1.id)).toEqual(distribution1);
```
Comment (decentralion) [inline src/core/ledger/ledger.test.js:976]: nit: may as well check for distribution 2.

**Review by decentralion — APPROVED** [review]
looks good. per `yarn coverage`, ledger continues to have 100% test coverage after this commit.

File: src/core/ledger/ledger.test.js (hunk 2 above)
Comment (blueridger) [reply src/core/ledger/ledger.test.js:976]: adding and merging

**Pull request merged**